### PR TITLE
[Feature] Admin Spam Sort

### PR DESCRIPTION
**Pull request by Ghalko**
## Purpose
To sort based on the newest report time rather than the oldest comment. ADM-63

## Repercussions
Migration.

### CODE REVIEW

File: admin/spam/views.py [hunk 1]
```diff
@@ -1,47 +1,149 @@
...
+from .forms import ConfirmForm
 
 
-def get_spam_list():
-    query = (
-        Q('reports', 'ne', {}) &
-        Q('reports', 'ne', None)
-    )
-    return Comment.find(query).sort('date_created')
+def key_order_comments(comment):
```
Comment (mfraezz) [inline admin/spam/views.py]: This appears to be the O(n^2) sorting mechanism mentioned in the PR description, but I don't see it being used anywhere. [This line](https://github.com/CenterForOpenScience/osf.io/pull/5045/files#diff-43b817f0cafd09bc0da4a5351442a2daR31) appears to be what's specifying the ordering.

Comment (mfraezz) [inline admin/spam/views.py]: Put a breakpoint in this function and it wasn't hit. Please remove.

Also, confirmed that it was sorting by most recent report first.

File: scripts/migrate_spam.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        date = None
+        for user, report in comment.reports.iteritems():
+            report_date = report.get('date')
+            if date is None or report_date > date:
+                date = report_date
+        comment.latest_report = date
+    comment.save()
+
+
+def main():
```
Comment (mfraezz) [inline scripts/migrate_spam.py]: Also confirmed that switching between branches swapped sort orderings without breaking anything -- this script may be superfluous.

File: scripts/migrate_spam.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    comments = Comment.find()
```
Comment (mfraezz) [inline scripts/migrate_spam.py]: `Comment.find(Q('spam_status', 'ne', 0))` should shave some time off this script, as well as obviate some of the mess in `migrate_comment`.

File: scripts/migrate_spam.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        if len(comment.reports) > 0:
+            comment.spam_status = Comment.FLAGGED
+    if comment.spam_status not in (Comment.FLAGGED, Comment.SPAM,
+                                   Comment.UNKNOWN, Comment.HAM):
+        comment.spam_status = Comment.UNKNOWN
+    try:
+        temp = comment.latest_report
+    except:
+        date = None
+        for user, report in comment.reports.iteritems():
```
Comment (mfraezz) [inline scripts/migrate_spam.py]: `comment.latest_report = max([r.get('date', None) for u, r in list(comment.reports.iteritems()])`

File: scripts/migrate_spam.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+
+def main():
+    comments = Comment.find()
+    for comment in comments:
+        migrate_comment(comment)
+
+
+if __name__ == '__main__':
+    main()
```
Comment (mfraezz) [inline scripts/migrate_spam.py]: Please use a `TokuTransaction`, add a file logger, and allow for a `dry-run`. See the [devdocs](http://cosdev.readthedocs.org/en/latest/osf/migrations.html?highlight=script#migrations) and [this part of this script](https://github.com/CenterForOpenScience/osf.io/blob/develop/scripts/dropbox/migrate_to_external_accounts.py#L255-L261) for reference

**Comment by mfraezz** [comment]
Please add a few tests, both for the script and for spam ordering.

**Comment by mfraezz** [comment]
Pass complete. A few small changes and this should be RTM :cactus:

File: admin/spam/views.py (hunk 1 above)
Comment (Ghalko) [inline admin/spam/views.py]: Removed unused function.

**Comment by Ghalko** [comment]
I guess I need a bit of a tutorial on modm. What I glean from this is that it loads the instance and anything that is empty gets filled with `None`?

**Comment by mfraezz** [comment]
`Comment.find()` returns a `MongoQuerySet` that isn't evaluated (i.e. models aren't loaded into memory) until it's iterated through, either via `list(query_result)` or by a `for` loop, list comprehension, etc. `.find` accepts a single `modularodm` query (or, `Q`) which limits the number of results based on ones that satisfy the query. `Q`s accept three parameters: `attribute`, `operator` (`ne` is not equal, `eq` is equals, `contains` or `icontains` perform regexs, etc), and `argument`, which can be anything (`boolean`, `int`, `str`, some form of complicated object).  Note: multiple `Q`'s `&`'d together count as a single `Q`. You can think of MODM queries, in this case, as a thin wrapper for interacting with the database -- significantly more thin than interacting with the model instances themselves.

So, `Comment.find(Q('spam_status', 'ne', 0) & Q('latest_report', 'eq', None))` limits the results to comments with a non-zero `spam_status` and no `latest_report`, which I think is the set of `Comment` objects the script should be interacting with.

**Comment by Ghalko** [comment]
Please merge #5013 before this one can proceed.

File: admin_tests/spam/test_views.py [hunk 6]
```diff
@@ -2,9 +2,49 @@
...
+        self.comment_3.report_abuse(user=self.user_2, save=True,
+                                    category='spam')
+        self.comment_4.report_abuse(user=self.user_2, save=True,
+                                    category='spam')
+        self.comment_5.report_abuse(user=self.user_1, save=True,
+                                    category='spam')
+        self.comment_6.report_abuse(user=self.user_1, save=True,
+                                    category='spam')
+
+    def test_get_spam(self):
```
Comment (mfraezz) [inline admin_tests/spam/test_views.py]: Should this test check for correct ordering?

File: scripts/migrate_spam.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+                date = report.get('date')
+            elif date < report.get('date'):
+                date = report.get('date')
+        record.latest_report = date
+        record.save()
+        logger.info('Migrated latest_report for comment {}'.format(record._id))
+
+
+def get_no_status_targets():
+    return [c for c in Comment.find(Q('spam_status', 'eq', None))]
```
Comment (mfraezz) [inline scripts/migrate_spam.py]: You can just return `Comment.find(query)` to save evaluation time.

File: scripts/migrate_spam.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        with TokuTransaction():
+            migrate_status(get_no_status_targets())
+            if dry_run:
+                raise Exception('Dry Run -- Aborting Transaction')
+    if latest:
+        with TokuTransaction():
+            migrate_latest(get_no_latest_targets())
+            if dry_run:
+                raise Exception('Dry Run -- Aborting Transaction')
+    init_app(set_backends=True, routes=False)
```
Comment (mfraezz) [inline scripts/migrate_spam.py]: You should be doing this before this migration methods are run

File: scripts/migrate_spam.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    if status:
+        with TokuTransaction():
+            migrate_status(get_no_status_targets())
+            if dry_run:
+                raise Exception('Dry Run -- Aborting Transaction')
+    if latest:
+        with TokuTransaction():
+            migrate_latest(get_no_latest_targets())
+            if dry_run:
+                raise Exception('Dry Run -- Aborting Transaction')
```
Comment (mfraezz) [inline scripts/migrate_spam.py]: You can get both migration methods to run during a dry run by changing this to something like:

```python
with TokuTransaction():
    if status:
        migrate_status(get_no_status_targets())
    if latest:
        migrate_latest(get_no_latest_targets())
    if dry_run:
        raise Exception('Dry Run -- Aborting Transaction')
```

File: scripts/migrate_spam.py (hunk 7 above)
Comment (Ghalko) [inline scripts/migrate_spam.py]: Done with previous vestiges.

File: scripts/migrate_spam.py (hunk 8 above)
Comment (Ghalko) [inline scripts/migrate_spam.py]: Done.

File: scripts/migrate_spam.py (hunk 9 above)
Comment (Ghalko) [inline scripts/migrate_spam.py]: Done.

File: admin_tests/spam/test_views.py (hunk 6 above)
Comment (Ghalko) [inline admin_tests/spam/test_views.py]: Checked.

File: scripts/migrate_spam.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    latest = True
+    if '--dry' in sys.argv:
+        dry_run = True
+    if not dry_run:
+        script_utils.add_file_logger(logger, __file__)
+    if '--no_status' in sys.argv:
+        status = False
+    if '--no_latest' in sys.argv:
+        latest = False
+    if not dry_run:
```
Comment (mfraezz) [inline scripts/migrate_spam.py]: This conditional is doubled

Comment (Ghalko) [inline scripts/migrate_spam.py]: This is done.

File: scripts/migrate_spam.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+def migrate_status(records):
+    for record in records:
+        if len(record.reports) > 0:
+            record.spam_status = Comment.FLAGGED
+        else:
+            record.spam_status = Comment.UNKNOWN
```
Comment (sloria) [inline scripts/migrate_spam.py]: Is there a difference between `spam_status=UNKNOWN` vs `spam_status=None`? Would it be better if `Comment.UNKNOWN=None`?

File: website/project/spam/model.py [hunk 12]
```diff
@@ -39,6 +39,7 @@ class SpamMixin(StoredObject):
     HAM = 4
 
     spam_status = fields.IntegerField(default=UNKNOWN, index=True)
+    latest_report = fields.DateTimeField(default=None, index=True)
```
Comment (sloria) [inline website/project/spam/model.py]: `latest_report` is a bit of a misnomer (it's not a report). For consistency with other datetime fields, how about `date_last_reported`?

**Comment by sloria** [comment]
Pass finished. Approach looks good. Just a few minor changes suggested above

File: website/project/spam/model.py (hunk 12 above)
Comment (Ghalko) [inline website/project/spam/model.py]: Done.

File: scripts/migrate_spam.py (hunk 11 above)
Comment (Ghalko) [inline scripts/migrate_spam.py]: This makes sense. I will reverse the migration to catch any current `UNKNOWN`s.

**Comment by sloria** [comment]
Looks good 🐙 . Merged into release/0.68.0

**Pull request closed without merging**